### PR TITLE
Course Requisites and Incompatibilities

### PR DESCRIPTION
**Pull request by spencerj411**
Aims to complete #81 and #78.
This pr contains the following additions/modification:

- added `rules` field to CourseModel
- added either/or rule to course
- added requisite/incompatibility section in the form
- added program field to requisite
- added incompatibility option
- added year-level field to subject-area units to satisfy all criteria (eg. 12 units of COMP 4000-level courses)

Note:
- The View feature will unfortunately have to be updated. #94 

**Whilst creating a course**
image

**Once it has been saved**
image

### CODE REVIEW

File: cassdegrees/templates/createcourse.html [hunk 1]
```diff
@@ -42,31 +41,47 @@
...
-
-            <p>
-                <label for="text">Incompatible Course(s)</label>
-                <select class="text tfull" name="incompatibles[]" size="5" multiple>
-                    {% for course in courses %}
-                        <option value="{{ course.code }}">{{ course.code }}</option>
-                    {% endfor %}
-                </select>
-            </p>
+{#            <!-- TODO: Unhandled so far -->#}
```
Comment (j-selby) [inline cassdegrees/templates/createcourse.html]: Wait, can this be removed now?

File: cassdegrees/templates/widgets/rulescripts.html [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                {{ component_names[details.type] }}
+            </p>
+            <input type="button" v-on:click="$emit('remove')" class="btn-uni-grad btn-snall no-left-margin" value="Remove" />
+        </header>
+        <div class="box-solid">
+            <div class="card-content" v-bind:is="'rule_' + details.type" v-bind:details="details"></div>
+        </div>
+    </div>
+</script>
+{% endverbatim %}
```
Comment (j-selby) [inline cassdegrees/templates/widgets/rulescripts.html]: Newline

**Review by j-selby — APPROVED** [review]
Nice work!

**Review by JackSekoranja — CHANGES_REQUESTED** [review]
I did find a few things that we might want to change:
- 'Course' and 'Incompatibility' should work the same. Maybe remove the unit count from 'Course' and re-name it 'Course Requisite'
- It says "Students must be studying a program which includes:", but you can only select one option
- There is a unit count on custom-text items. We might not want this since it may be things like "Students need to get a permission code for this course"
- Units is also mentioned in other places (Subject-area, Level-specific), but it may not make sense in this context

Screen Shot 2019-05-22 at 2 00 11 pm

Otherwise it looks very good, well done :)

**Comment by Daniel-Jang** [comment]
@JackSekoranja I had a look at your comment and the first 3 dot points I agree with. However, with your 4th point, there are cases where a course requisite would be like "... and 6 units of 1000 level MATH":
https://programsandcourses.anu.edu.au/2019/course/COMP2100

I will fix everything else, however, since you raised valid points about them.

Edit: Completed the first 3 suggestions, ready for review.

**Review by JackSekoranja — APPROVED** [review]
Looks good

**Pull request merged**